### PR TITLE
FIX: preserve original keys after key_builder transformation

### PR DESCRIPTION
**Pull request by haizhou**
## What do these changes do?

Preserve original keys with the underlying function call after `key_builder` transformation takes place in `multi_cached`

## Are there changes in behavior for the user?

There shouldn't be if the intention is to allow differently formatted cache keys than the original keys passed in.

## Related issue number

https://github.com/aio-libs/aiocache/issues/563

## Checklist

- [ ] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` (e.g. `588.bugfix`)
  * if you don't have an `issue_id` change it to the pr id after creating the PR
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: `Fix issue with non-ascii contents in doctest text files.`

### CODE REVIEW

**Comment by Dreamsorcerer** [comment]
Looks right to me, but can you add a test to demonstrate the problem (and merge master).

Sorry for the delay.

File: aiocache/decorators.py [hunk 1]
```diff
@@ -310,19 +309,19 @@ async def decorator(
...
-                    missing_keys.append(key)
+                    missing_keys.append(orig_key)
                 else:
-                    partial[key] = value
+                    partial[orig_key] = value
             if values and None not in values:
                 return partial
         else:
-            missing_keys = list(keys)
+            missing_keys = list(orig_keys)
```
Comment (pshafer-als) [inline aiocache/decorators.py:324]: I think you're right that the keys from `keys_from_attr` should get stored in `missing keys`, because those get passed to the callable.

File: aiocache/decorators.py [hunk 2]
```diff
@@ -310,19 +309,19 @@ async def decorator(
...
-            values = await self.get_from_cache(*keys)
-            for key, value in zip(keys, values):
+            values = await self.get_from_cache(*cache_keys)
+            for orig_key, value in zip(orig_keys, values):
                 if value is None:
-                    missing_keys.append(key)
+                    missing_keys.append(orig_key)
                 else:
-                    partial[key] = value
+                    partial[orig_key] = value
```
Comment (pshafer-als) [inline aiocache/decorators.py:320]: No, I think that `partial` should use the mapped keys returned by `key_builder`. These keys need to match the keys in the returned data, not the keys passed into the callable.

File: aiocache/decorators.py [hunk 3]
```diff
@@ -310,19 +309,19 @@ async def decorator(
...
+        orig_keys, cache_keys, new_args, args_index = self.get_cache_keys(f, args, kwargs)
 
         if cache_read:
-            values = await self.get_from_cache(*keys)
-            for key, value in zip(keys, values):
+            values = await self.get_from_cache(*cache_keys)
+            for orig_key, value in zip(orig_keys, values):
                 if value is None:
-                    missing_keys.append(key)
+                    missing_keys.append(orig_key)
```
Comment (pshafer-als) [inline aiocache/decorators.py:318]: Agreed.  See the comment below regarding `missing_keys`.

File: aiocache/decorators.py [hunk 4]
```diff
@@ -310,19 +309,19 @@ async def decorator(
     ):
         missing_keys = []
         partial = {}
-        keys, new_args, args_index = self.get_cache_keys(f, args, kwargs)
+        orig_keys, cache_keys, new_args, args_index = self.get_cache_keys(f, args, kwargs)
```
Comment (pshafer-als) [inline aiocache/decorators.py:312]: Consider naming these as `in_keys` and `out_keys` for clarity.  Respectively they are the keys that are passed in to the callable, and the keys that are returned by the callable (`out_keys` are also the translated keys returned by `key_builder`).

If these `in_keys` are only used internally, then it might make sense to leave `get_cached_keys()` unmodified (i.e., revert it).  Then instead `in_keys` could be generated here from `_get_args_dict()`. `get_cache_keys()` is a public member function.

File: aiocache/decorators.py (hunk 2 above)
Comment (pshafer-als) [reply aiocache/decorators.py:320]: > No, I think that `partial` should use the mapped keys returned by `key_builder`. These keys need to match the keys in the returned data, not the keys passed into the callable.

This is under the assumption of [Interpretation 2](https://github.com/aio-libs/aiocache/pull/564#issuecomment-1380574858).

File: aiocache/decorators.py (hunk 4 above)
Comment (pshafer-als) [reply aiocache/decorators.py:312]: > Consider naming these as `in_keys` and `out_keys` for clarity. Respectively they are the keys that are passed in to the callable, and the keys that are returned by the callable (`out_keys` are also the translated keys returned by `key_builder`).

This is under the assumption of [Interpretation 2](https://github.com/aio-libs/aiocache/pull/564#issuecomment-1380574858).

**Comment by haizhou** [comment]
Agree with @pshafer-als on need for clarification about intended use case for `key_builder` in `multi_cached`. I added an example, but as of the latest release, it is raising a KeyError on https://github.com/aio-libs/aiocache/blob/0.11.1/aiocache/decorators.py#L366 because the key returned from the result has the already transformed key by `key_builder` but is passed through again

**Comment by Dreamsorcerer** [comment]
> Hmmm, there still seems to be some ambiguity regarding the dict keys that are returned by the decorated function and the role of `key_builder` in `multi_cached`. I see two interpretations. Can someone help to clarify the intended usage (or the prominent use case)?

If I'm understanding correctly, you're suggesting in interpretation 2, that a function can return keys which don't match the input? Like:
```
def foo(keys=(1, 2)):
    return {"a": "foo", "b": "bar"}
```

That's not my understanding of the documentation.

> By default, the callable’s name and call signature are not incorporated into the cache key, so if there is another cached function returning a dict with same keys, those keys will be overwritten. To avoid this, use a specific namespace in each cache decorator or pass a key_builder.

So, the primary purpose is to avoid cache key collisions with other functions. Not so sure why you'd use it over namespace though...

> the values of keys_from_attr will be transformed before requesting them from the cache.

So, we use key_builder for the key to request from cache.

> Equivalently, the keys in the dict-like mapping returned by the decorated callable will be transformed before storing them in the cache.

And we use key_builder on the returned keys to set them in the cache.

In other words, it seems like key_builder just transforms keys to deal with the cache. The function should have no knowledge of the caching details, and the transformed keys should not be visible outside of the caching code.

Which means something like this should work:
```
def my_func(*args):
    ...

f1 = multi_cached("args", key_builder=builder1)(my_func)
f2 = multi_cached("args", key_builder=builder2)(my_func)
my_func(1)
f1(1)
f2(1)
# All 3 calls should return the same result, with different caches used for the latter 2 functions.
```

**Comment by Dreamsorcerer** [comment]
So, I think the changes here are still correct. If we can get a new test for this (in tests/ut/test_decorators.py using a mock cache should work), then we can merge this.

File: tests/ut/test_decorators.py [hunk 5]
```diff
@@ -590,6 +590,13 @@ async def bar():
 
         assert foo.cache != bar.cache
 
+    async def test_key_builder(self):
+        @multi_cached("keys", key_builder=lambda key, _, keys: key + 1)
+        async def incr(keys=None):
+            return {k: k + 1 for k in keys}
+        
+        assert incr(keys=[1]) == {1: 3}
```
Comment (Dreamsorcerer) [inline tests/ut/test_decorators.py:598]:
```suggestion
        @multi_cached("keys", key_builder=lambda key, _, keys: key + 1)
        async def incr(keys=None):
            return {k: k * 3 for k in keys}
        
        assert incr(keys=(1,)) == {1: 3}
        cached_value = await incr.cache.get(2)
        assert cached_value == 3
        assert not await incr.cache.exists(1)
```

Maybe something like this is more comprehensive.

File: examples/multicached_decorator.py [hunk 6]
```diff
@@ -39,6 +49,8 @@ async def test_multi_cached():
     await cache.delete("d")
     await cache.close()
 
+    await multi_cached_key_builders(keys=(1, 2)) == {1: "Z", 2: "Y"}
```
Comment (Dreamsorcerer) [inline examples/multicached_decorator.py:52]: I'm not sure we need more in the example, so I'd suggest reverting the changes to this file (the code is wrong currently anyway).

File: tests/ut/test_decorators.py [hunk 7]
```diff
@@ -590,6 +590,16 @@ async def bar():
...
 
+    async def test_key_builder(self):
+        @multi_cached("keys", key_builder=lambda key, _, keys: key + 1)
+        async def incr(keys=None):
+            return {k: k * 3 for k in keys}
+        
+        assert await incr(keys=(1,)) == {1: 3}
+        cached_value = await incr.cache.get(2)
+        assert cached_value == 3
+        assert not await incr.cache.exists(1)
```
Comment (Dreamsorcerer) [inline tests/ut/test_decorators.py:601]:
```suggestion
        @multi_cached("keys", key_builder=lambda key, _, keys: key + 1)
        async def f(keys=None):
            return {k: k * 3 for k in keys}


        assert await f(keys=(1,)) == {1: 3}
        cached_value = await f.cache.get(2)
        assert cached_value == 3
        assert not await f.cache.exists(1)
```

Comment (Dreamsorcerer) [reply tests/ut/test_decorators.py:601]: Huh, the commit button is not there...

Comment (Dreamsorcerer) [reply tests/ut/test_decorators.py:601]: Oh, you must not have allowed maintainers write access...

**Comment by Dreamsorcerer** [comment]
@haizhou There are 6 failing tests, due to the change of return value in .get_cache_keys(). Can you enable write access to you branch or update those tests?

All the failing tests are under tests/ut/test_decorators.py::TestMultiCached

File: tests/ut/test_decorators.py [hunk 8]
```diff
@@ -590,6 +590,17 @@ async def bar():
 
         assert foo.cache != bar.cache
 
+    async def test_key_builder(self):
+        @multi_cached("keys", key_builder=lambda key, _, keys: key + 1)
+        async def f(keys=None):
+            return {k: k * 3 for k in keys}
+
+
```
Comment (Dreamsorcerer) [inline tests/ut/test_decorators.py:598]:
```suggestion
```

I also misread the flake8 error.

**Comment by haizhou** [comment]
> @haizhou There are 6 failing tests, due to the change of return value in .get_cache_keys(). Can you enable write access to your branch or update those tests?
> 
> All the failing tests are under tests/ut/test_decorators.py::TestMultiCached

@Dreamsorcerer thanks for the collaboration. I'm not sure if we have enough seats left for adding an outside collaborator to our repos, but I took an attempt at fixing the tests

**Comment by Dreamsorcerer** [comment]
> > @haizhou There are 6 failing tests, due to the change of return value in .get_cache_keys(). Can you enable write access to your branch or update those tests?
> > All the failing tests are under tests/ut/test_decorators.py::TestMultiCached
> 
> @Dreamsorcerer thanks for the collaboration. I'm not sure if we have enough seats left for adding an outside collaborator to our repos, but I made an attempt to fix the tests

No problem. I think there's usually a checkbox to enable maintainer write access when opening a PR (not sure how to find it after opening though). That shouldn't affect those seats (and probably only applies to the branch, rather than the entire repo).

File: tests/ut/test_decorators.py [hunk 9]
```diff
@@ -395,26 +395,27 @@ def test_alias_takes_precedence(self, mock_cache):
...
     def test_get_cache_keys_missing_kwarg(self, decorator):
-        assert decorator.get_cache_keys(stub_dict, (), {}) == ([], [], -1)
+        assert decorator.get_cache_keys(stub_dict, (), {}) == ([], [], [], -1)
 
     def test_get_cache_keys_arg_key_from_attr(self, decorator):
         def fake(keys, a=1, b=2):
             """Dummy function."""
 
-        assert decorator.get_cache_keys(fake, (["a"]), {}) == (["a"], [["a"]], 0)
+        assert decorator.get_cache_keys(fake, (["a"]), {}) == ("a", ["a"], ["a"], 0)
```
Comment (Dreamsorcerer) [inline tests/ut/test_decorators.py:410]:
```suggestion
        assert decorator.get_cache_keys(fake, (["a"],), {}) == ("a", ["a"], ["a"], 0)
```

I think the expected output was correct before, someone just clearly messed up the input. The test passed before by luck, because `["a"][0] == "a"[0]`.
